alphagov/pay-ci
# Improve Version Regression Check

Allow earlier versions to be deployed if they are the latest enabled version. This allows developers to deploy earlier versions by disabling any later versions whilst protecting from unintended regression due to Concourse quirks.


### HOW TO TEST ###
Can be run locally with the following, adjusting env vars to prompt various behaviour. Log in first locally with `fly login --target pay-dev`

```
CONTAINER_SUFFIX=test PIPELINE_NAME=deploy-to-test FLY_USERNAME=pay-dev APP_NAME=adminusers CLUSTER_NAME=test-12-fargate APPLICATION_IMAGE_TAG=1919 aws-vault exec pay-test-admin -- node check-release-versions.js


You are trying to deploy release  1919 of adminusers which is older than the current release 1920.

Checking that 1919 is the latest ENABLED release.
The latest ENABLED release is 1920
1919 of adminusers is not the latest ENABLED version. Bailing out.
If you need to deploy this version you must disable later versions (pinning will not work)

```
and for example
```
➤➤ scripts $ CONTAINER_SUFFIX=test PIPELINE_NAME=deploy-to-test FLY_USERNAME=pay-dev APP_NAME=adminusers CLUSTER_NAME=test-12-fargate APPLICATION_IMAGE_TAG=1919 aws-vault exec pay-test-admin -- node check-release-versions.js


You are trying to deploy release  1919 of adminusers which is older than the current release 1920.

Checking that 1919 is the latest ENABLED release.
The latest ENABLED release is 1919
Proceeding to deploy release 1919
```


Please help by checking for typos, it seems a little brittle in the reliance on config values between the pipeline and the node task. I have tried to consider various failure scenarios with helpful error messages. This new logic should only be invoked when someone attempts to deploy an older version. There is retry logic because sometimes the fly check resources command does not return any results...
